RFQ-T docs

feuGeneA opened the pull request:
To view the rendering locally, edit `/etc/hosts` to alias 127.0.0.1 to `0xproject.localhost`, check out the repo, do `yarn`, then `dev`, then navigate to, eg, https://0xproject.localhost:3572/docs/guides/understanding-rfqt

feuGeneA committed:
Placeholders

feuGeneA committed:
Incomplete draft of guide to RFQ-T in 0x API

feuGeneA commented:
For convenience, here's a quick view of the rendering of the beginning of the main guide as it stands right now:
![image](https://user-images.githubusercontent.com/7883777/89209629-af375c00-d58c-11ea-85cd-8eb80ad86921.png)

feuGeneA commented on mdx/api/index.mdx:
```diff
@@ -682,6 +682,39 @@ Easily wrap and unwrap between ETH and WETH by requesting a swap quote by settin
 
+#### Ensure RFQ-T liquidity
+
+RFQ-T liquidity can only be accessed if the request supplies a taker address, a permissioned API key via header `0x-api-key`, and the query parameter `intentOnFilling=true`.  For more details see the guide [Understanding RFQ-T and the 0x API](/docs/guides/understanding-rfqt-0x-api).  Also, ensure that `Native` sources have not been excluded.
```
Need to also prescribe here the inclusion of a `takerAddress`.

feuGeneA replied:
Need to also prescribe here the inclusion of a non-null `takerAddress` query param in the request.

PirosB3 replied:
Just to check, do we need a "What is RFQT liquidity" chapter, or can we assume it somehow? my understanding is that this documentation will be rendered on the main webpage, therefore we may need an introduction.

PirosB3 replied:
The integrator needs to also ensure that `excludedSources` key does not contain `Native` or `0x`.

PirosB3 commented on mdx/api/index.mdx:
```diff
@@ -682,6 +682,39 @@ Easily wrap and unwrap between ETH and WETH by requesting a swap quote by settin
+
+RFQ-T liquidity can only be accessed if the request supplies a taker address, a permissioned API key via header `0x-api-key`, and the query parameter `intentOnFilling=true`.  For more details see the guide [Understanding RFQ-T and the 0x API](/docs/guides/understanding-rfqt-0x-api).  Also, ensure that `Native` sources have not been excluded.
+
+An RFQ-T quote can be identified by its inclusion of one or more orders with a `takerAddress` matching the `takerAddress` that was supplied in the request.  However, the lack of such an order in a quote does not indicate that RFQ-T liquidity was not considered; it only indicates that RFQ-T liquidity was not competitive enough to be included in the quote.
```
it's inclusion. Typo?

PirosB3 commented on mdx/api/index.mdx:
```diff
@@ -682,6 +682,39 @@ Easily wrap and unwrap between ETH and WETH by requesting a swap quote by settin
+
+An RFQ-T quote can be identified by its inclusion of one or more orders with a `takerAddress` matching the `takerAddress` that was supplied in the request.  However, the lack of such an order in a quote does not indicate that RFQ-T liquidity was not considered; it only indicates that RFQ-T liquidity was not competitive enough to be included in the quote.
+
+<HttpCall fontSize="16px" path="/swap/v0/quote?buyToken=WETH&sellToken=ETH&buyAmount=10000000&takerAddress=0xffffffffffffffffffffffffffffffffffffffff" />
```
What kind of request is this? selling `1e-11` WETH for ETH?

PirosB3 commented on mdx/guides/understanding-rfqt.mdx:
```diff
@@ -0,0 +1,62 @@
+
+With the use of RFQ-T, the API aggregates not just on-chain pools and Mesh order books, but also real-time quotes from live market makers offering exclusive pricing.
+
+RFQ is an abbreviation of Request for Quote, in that the 0x API requests quotes from market makers on behalf of its clients. The -T suffix indicates that the settlement transaction is to be submitted to the blockchain by the Taker. (An RFQ-M model, in which the maker submits the transaction, is under consideration, but is outside the scope of this Guide.)
```
💯

PirosB3 commented on mdx/guides/understanding-rfqt.mdx:
```diff
@@ -0,0 +1,62 @@
+
+## Trusted Takers
+
+The 0x API now automatically attempts to supplement every quote with RFQ-T liquidity. The biggest determinant of whether such live quotes will be aggregated with other liquidity sources is whether the client is a known and trusted taker
```
RFQT liquidity is only supplied to the integrator, if they have an agreement (technically speaking, an API key) with us.

PirosB3 commented on mdx/guides/understanding-rfqt.mdx:
```diff
@@ -0,0 +1,62 @@
+
+The 0x API now automatically attempts to supplement every quote with RFQ-T liquidity. The biggest determinant of whether such live quotes will be aggregated with other liquidity sources is whether the client is a known and trusted taker
+
+Trusted takers are identified by an API key, specified in the HTTP header `0x-api-key`. In addition, to qualify for RFQ-T liquidity, the request must also specify a non-null `takerAddress` query parameter, which will be included in the order that the maker signs, preventing anyone else from filling that order.
```
This is kind of repetitive and already covered in `mdx/api/index.mdx`, have you considered having 1 document link to the other? this way, we don't have to maintain content on both pages.

PirosB3 commented on mdx/guides/understanding-rfqt.mdx:
```diff
@@ -0,0 +1,62 @@
+
+In addition to the 0x API configuration identifying trusted takers, it also contains a list of specific market makers that participate in the RFQ-T model. Each maker is identified by an HTTP API endpoint URL, and each endpoint has an associated list of asset pairs for which that endpoint will provide quotes.
+
+For the instance at `api.0x.org`, the 0x team is maintaining a list of trusted market makers, which at this time includes 0x's own Periscope Trading, as well as a handful of other trusted partners.
```
It would be nice to link out to [this article](https://blog.0xproject.com/introducing-0x-labs-and-periscope-trading-c5f84879e782)

PirosB3 commented on mdx/guides/understanding-rfqt.mdx:
```diff
@@ -0,0 +1,62 @@
+
+To aleviate this potential capital inefficiency, the RFQ-T model introduces the distinction between "indicative pricing" and "firm quotes."
+
+The 0x API offers two different endpoints for the use of RFQ-T. One is an "indicative pricing" endpoint, hosted at `/swap/vX/price`. This endpoint responds with pricing information, but that response does not contain a full 0x order, so it does not constitute a legitimate transaction that can be submitted to the blockchain. Therefore, when a market maker services an indicative request they do not ...
```
I see what you are doing with `vX`, however, do you think it's worth specifying somewhere that `vX` is the latest version. Just to ensure 100% no ambiguity?

PirosB3 commented on mdx/guides/understanding-rfqt.mdx:
```diff
@@ -0,0 +1,62 @@
+
+Note that the `/swap/vX/quote` endpoint is also used to access non-RFQ-T liquidity as well. In order to qualify for access to RFQ-T liquidity, the request must include query parameter `intentOnFilling=true` (in addition to the aforementioned `0x-api-key` and non-null `takerAddress`).
+
+0x API clients should strictly adhere to a rule of using indicative pricing by default, and of only requesting firm quotes when the user is actually intent on filling. For example, a continuously updating price display on a web page should be fed by indicative pricing, and a trade confirmation display should be fed by firm quotes.
```
great examples! 💯

PirosB3 reviewed (changes requested):
@blakehendo should we also mention the new `includedSources` parameter?

Gene the doc is a great start, I always admire your way of describing things very easily with simple words. Let me know when I should do a second review

feuGeneA committed:
Complete draft of guide to RFQ-T in 0x API

feuGeneA committed:
scripts: clarify need for API key when deploying

feuGeneA committed:
fixup complete draft

feuGeneA committed:
docs: remove duplicate '0x API' tag from FAQ Guide

feuGeneA commented:
I've pushed a complete draft of the main Guide.  It's currently deployed to staging, accessible at http://staging-0xproject.s3-website-us-east-1.amazonaws.com/docs/guides/rfqt-in-the-0x-api

feuGeneA committed:
Fix sitemap after recent Guide rename

feuGeneA committed:
RFQ-T Guide: miscellaneous revisions

- endpoints/paths => resources
- users => clients
- added links
- say makers MUST return a 204 to decline to quote
- prettier
- clarified maker and taker roles in parameter documentation

feuGeneA committed:
API Spec: Document RFQ-T specifics

feuGeneA committed:
MetaTx Guide: Tweak indicative quote example

Tweaked to stop advising that people submit the indicative quote to the
blockchain, which doesn't make any sense.

feuGeneA committed:
API Spec: Document /swap/quote's intentOnFilling

feuGeneA committed:
API Spec: Update /swap/quote RFQT example

feuGeneA committed:
API spec: Clarify Native vs 0x source exclusion

feuGeneA committed:
RFQ-T Guide: Tweak wording, add links

PirosB3 commented on mdx/api/index.mdx:
```diff
@@ -171,6 +171,8 @@ When both are provided, `sellAmount` takes precedence over `buyAmount`.
 | `gasPrice`           | (Optional, defaults to ethgasstation "fast") The target gas price (in wei) for the swap transaction. If the price is too low to achieve the quote, an error will be returned.                                       |
 | `takerAddress`       | (Optional) The address which will fill the quote. When provided the gas will be estimated and returned. An eth_call will also be performed. If this fails a Revert Error will be returned in the response.          |
 | `excludedSources`    | (Optional) Liquidity sources (`Eth2Dai`, `Uniswap`, `Kyber`, `0x`, `LiquidityProvider` etc) that will not be included in the provided quote. Ex: `excludedSources=Uniswap,Kyber,Eth2Dai`. See [here](https://github.com/0xProject/0x-monorepo/blob/development/packages/asset-swapper/src/utils/market_operation_utils/types.ts#L27) for a full list of sources  |
+| `skipValidation`     | (Optional) Normally, whenever a `takerAddress` is provided, the API will validate the quote for the user. (For more details, see ["How does `takerAddress` help with catching issues?"](/docs/guides/0x-api-faq#how-does-takeraddress-help-with-catching-issues).)  When this parameter is set to `true`, that validation will be skipped. See also [here](/docs/guides/rfqt-in-the-0x ...
```
Hey Gene, so I think me and you need to sync about RFQT params. I still have [this PR](https://github.com/0xProject/website/pull/318/files) open that I was wanting to merge when RFQ is out. Do you think it's worth merging mine in yours?

PirosB3 commented on mdx/guides/unlocking-meta-transaction-swaps.mdx:
```diff
@@ -31,8 +31,6 @@ import * as qs from 'qs';
         `https://api.0x.org/meta_transaction/v0/price?${qs.stringify(params)}`
     );
- 
-    await window.web3.eth.sendTransaction(await response.json());
```
was this removed by mistake?

PirosB3 reviewed (approved):
This is great. I have a few questions:

- Are we not going to be covering the V1 endpoints?
- It would be nice to merge my small PR into your PR (I suggested that in the comments)

feuGeneA committed:
Market Making Guide: minor copy edits

feuGeneA committed:
Market Making Guide: add Mesh mentions

feuGeneA committed:
Market Making Guide: Add RFQ-T mentions

feuGeneA committed:
package.json: stop tool chains upon failures

feuGeneA committed:
API Spec: add validation codes & swap param

Changes ported from https://github.com/0xProject/website/pull/318

at the request of @PirosB3 in comment https://github.com/0xProject/website/pull/322#discussion_r469321388

steveklebanoff commented on mdx/api/index.mdx:
```diff
@@ -682,6 +688,115 @@ Easily wrap and unwrap between ETH and WETH by requesting a swap quote by settin
+
+RFQ-T liquidity can only be accessed if the request supplies a taker address, a permissioned API key via header `0x-api-key`, and the query parameter `intentOnFilling=true`. For more details see the guide [Understanding RFQ-T and the 0x API](/docs/guides/rfqt-in-the-0x-api).
+
+<HttpCall fontSize="16px" path="/swap/v0/quote?sellToken=WETH&buyToken=DAI&sellAmount=1000000000000000000&takerAddress=0xffffffffffffffffffffffffffffffffffffffff&intentOnFilling=true" />
```
![image](https://user-images.githubusercontent.com/210914/90185651-e3b8ce00-dd6b-11ea-8d2b-0c620d2f20e0.png)

This falls off out of the grey box - can we make the font smaller or break into multiple lines?

steveklebanoff commented on mdx/api/index.mdx:
```diff
@@ -837,6 +952,84 @@ Response will be an array of token symbols and the respective price in `sellToke
+
+### Response
+
+Identical to the response schema for `/swap/v0/quote`, with one exception: the `orders` property will always be undefined.
```
nit, but would prefer to say there is no `orders` key on the object returned.

the way this is written makes me feel like I would expect an `orders` key to be in the object, and it explicitly set to `undefined`

steveklebanoff commented on mdx/api/index.mdx:
```diff
@@ -837,6 +952,84 @@ Response will be an array of token symbols and the respective price in `sellToke
+
+#### Get the price available for selling 1 ETH for DAI
+
+<HttpCall path="/swap/v0/price?sellToken=WETH&buyToken=DAI&sellAmount=1000000000000000000" />
```
I think we should include a taker address in the example since this is expected when using RFQT

steveklebanoff commented on mdx/api/index.mdx:
```diff
@@ -837,6 +952,84 @@ Response will be an array of token symbols and the respective price in `sellToke
+
+### Request
+
+Identical to the request schema for `/swap/v0/quote`, with one exception: the `skipValidation` parameter will always be considered to be `true`, even when a `takerAddress` has been specified.
```
Should we note that they should send in a takerAddress if possible for the most accurate RFQT pricing?

steveklebanoff commented on mdx/api/index.mdx:
```diff
@@ -860,7 +1053,7 @@ When both are provided, `sellAmount` takes precedence over `buyAmount`.
 | `buyAmount`          | (Optional) The amount of `buyToken` (in `buyToken` base units) you want to receive.                                                                                                                                      |
 | `slippagePercentage` | (Optional) The maximum acceptable slippage in % of the `buyToken` amount if `sellAmount` is provided, the maximum acceptable slippage in % of the `sellAmount` amount if `buyAmount` is provided. This parameter will change over time with market conditions. |
-| `excludedSources`    | (Optional) Liquidity sources (`Eth2Dai`, `Uniswap`, `Kyber`, `0x`, `LiquidityProvider` etc) that will not be included in the provided quote. Ex: `excludedSources=Uniswap,Kyber,Eth2Dai`. See [here](https://github.com/0xProject/0x-monorepo/blob/development/packages/asset-swapper/src/utils/market_operation_utils/types.ts#L27) for a full list of sources  |
+| `excludedSources`    | (Optional) Liquidity sources (`Eth2Dai`, `Uniswap`, `Kyber`, `0x`, `LiquidityProvider` etc) that will not be included in the provided quote. Ex: `excludedSources=Uniswap,Kyber,Eth2Dai`. See [here](https://github.com/0xProject/0x-monorepo/blob/development/packages/asset-swapper/src/utils/market_operation_utils/types.ts#L27) for a full list of sources, but note that for thi ...
```
If we are going to do this call-out of Native sources here, should we do it everywhere we describe `excludedSources`?

steveklebanoff commented on mdx/guides/market-making-on-0x.mdx:
```diff
@@ -1,24 +1,26 @@
+0x orders can be hosted by anyone, rather than there being a single entity hosting the orders. Any entity that hosts and distributes 0x orders is called a "relayer." Currently most relayers are centralized entities hosting an API from which orders can be retrieved.
 
-A 0x order is any data packet that contains an [order's details](https://0x.org/docs/guides/v3-specification#order-message-format) together with a [valid signature](https://0x.org/docs/guides/v3-specification#signature-types) (e.g. a valid elliptic curve signature generated from the traders Ethereum address) of it's contents.
+In addition to relayer hosting, distribution can also be achieved via Mesh, a peer-to-peer network built specifically for sharing 0x orders. For more information, see [0x.org/mesh](0x.org/mesh).
```
Nice addition 👍

steveklebanoff commented on mdx/guides/market-making-on-0x.mdx:
```diff
@@ -42,14 +44,18 @@ In both models, the relayer never has custody over a trader's assets.
 
 If you want to be notified whenever the order is filled by a trader, you should also submit the order to a [0x Mesh node](https://0x-org.gitbook.io/mesh/) instance you are running.
 
+Whereas posting orders, to a relayer or to Mesh, represents a "push" distribution model, there is an alternative "pull" model as well, known as Request for Quote (RFQ).  This is a feature of the [0x API](/docs/api), which implements not just the Standard Relayer API, but also additional resources for aggregating liquidity from both 0x and non-0x sources. For RFQ, the 0x API is configured with a l ...
```
Nice 👍

steveklebanoff commented on mdx/guides/market-making-on-0x.mdx:
```diff
@@ -92,7 +98,7 @@ There are multiple ways to cancel 0x orders by interacting with the [Exchange co
 The `cancelOrdersUpTo` approach is perhaps the least straight forward but also the most powerful for a market maker. It allows for the cancellation of an arbitrary number of orders for a fixed amount of gas. By creating orders where the `salt` field is equal to the current unix timestamp in milliseconds, you can cancel all orders that were created at or before a certain time in a single `cancelOr ...
 
-Explicitly cancelling orders always requires an on-chain transaction. However, you may create short lived orders and replace the expired orders without any on-chain transactions. Due to the variability of transaction inclusion times in blocks, it is not recommended to set extremely aggressive expiration times for orders (empirically, most market makers set order expirations to ~5 minutes). Many r ...
+Explicitly cancelling orders always requires an on-chain transaction. However, you may create short lived orders and replace the expired orders without any on-chain transactions. Due to the variability of transaction inclusion times in blocks, it is not recommended to set extremely aggressive expiration times for orders. (Empirically, most market makers set order expirations to about 5 minutes.)  ...
```
I feel like this 5 minute number should be updated, perhaps we can ask @alexkroeger if we have any stats we could pull here

steveklebanoff commented on mdx/guides/rfqt-in-the-0x-api.mdx:
```diff
@@ -0,0 +1,164 @@
+
+Market makers providing open liquidity are subject to unique risks that impact their bottom-line pricing. If asset prices exhibit great movement in the markets, outstanding orders can flip from profitable to loss-incurring. And cancelling outstanding orders can be expensive and unreliable, subject to gas markets and front-running. Carrying these risks has a real cost, and that cost effectively de ...
+
+In practice, such losses are largely at the hands of one particular class of takers: arb bots. If makers share orders with an unspecified/null taker address, as is typical in Mesh order sharing, then those orders can be filled by anyone, including arb bots. Human traders, on the other hand, are extremely unlikely to be able to take advantage of such arbitrage opportunities. If market makers could ...
```
Suggestion: `arb` -> `arbitrage`

steveklebanoff commented on mdx/guides/rfqt-in-the-0x-api.mdx:
```diff
@@ -0,0 +1,164 @@
+
+Furthermore, the request sent from the 0x API to an RFQ-T market maker also includes the API key associated with the trusted integration client. In the same way that makers can blacklist taker addresses, they also have the option to blacklist specific applications that integrate with the 0x API.
+
+In order to retain access to RFQ-T liquidity, it's important for clients of the 0x API to be good citizens of the marketplace. For end users of trusted integrations, being a good citizen means actually filling the orders they claim to intend to fill. And for trusted integrations, it means properly managing their users' intentions; more specifically, it means adhering to the flow of using indicati ...
```
Feels a little weird to use "indicative" and "firm" quotes here without first introducing what they are. That being said, we introduce them immediately afterwards, so seems OK 👍

steveklebanoff commented on mdx/guides/rfqt-in-the-0x-api.mdx:
```diff
@@ -0,0 +1,164 @@
+
+In order to improve capital efficiency, the RFQ-T model introduces a new request/response flow: first request pricing, then request a fillable quote.
+
+When a market maker provides a fillable quote, which can be settled by the taker's submitting it to the 0x protocol, the taker has an expectation that the swap will actually settle, that it won't revert due to a lack of available maker assets. To uphold this expectation, makers will refrain from using those same assets as the basis for any other quotes, until the quote's expiry.
```
I don't think all market makers are doing this exactly as described, they are likely taking an educated guess on the likelihood that the user will actually fill, and not 100% reserving funds for every firm quote they give out.

steveklebanoff commented on mdx/guides/rfqt-in-the-0x-api.mdx:
```diff
@@ -0,0 +1,164 @@
+
+## Firm Quotes
+
+The firm quote resource is hosted at [`/swap/v0/quote`](/docs/api#get-swapv0quote) and responds with a full 0x order, which can be submitted to an Ethereum node by the client. Therefore it is expected that the maker has reserved the maker assets required to settle the trade, leaving the order unlikely to revert.
```
Referring to "full 0x order" seems a bit confusing as /quote contains calldata that is made up of potentially many orders. Perhaps we say something like "responds with on-chain fillable calldata"

steveklebanoff replied:
Also -- not 100% necessary -- but perhaps we start off with something like "Requesting a firm quote utilizes the existing the /swap/v0/quote with some additional required parameters"

steveklebanoff commented on mdx/guides/rfqt-in-the-0x-api.mdx:
```diff
@@ -0,0 +1,164 @@
+
+## When to Request Firm Quotes
+
+0x API clients should strictly adhere to a rule of using indicative pricing by default, and of only requesting firm quotes when the user is actually intent on filling. For example, a continuously updating price display on a web page should be fed by indicative pricing, and a trade confirmation display should be fed by firm quotes.
```
Great description 👍

steveklebanoff commented on mdx/guides/rfqt-in-the-0x-api.mdx:
```diff
@@ -0,0 +1,164 @@
+The 0x API will decline to relay from maker to taker any orders that do not meet the following criteria:
+
+-   `takerAddress` matches what was sent in the request
+-   `takerAssetAmount` equal to or less than the `takerAssetAmount` given in the request.
```
We actually accept if it is more -- some MMers do this to avoid rounding issues

steveklebanoff reviewed (commented):
A few comments but overall looks good to go -- great work!

feuGeneA committed:
RFQ-T Guide: Update diagram
